src/data-pipeline/lambda/emr-serverless-app/index.ts

```diff
@@ -0,0 +1,157 @@
+import { CloudFormationCustomResourceEvent, Context } from 'aws-lambda';
+import { logger } from '../../../common/powertools';
+import { putStringToS3, readS3ObjectAsJson } from '../../../common/s3';
+import { aws_sdk_client_common_config } from '../../../common/sdk-client-config';
+
+export interface ResourcePropertiesType {
+  ServiceToken: string;
+  projectId: string;
+  name: string;
+  version: string;
+  secourityGroupId: string;
+  subnetIds: string;
+  idleTimeoutMinutes: string;
+}
+
+const region = process.env.AWS_REGION!;
+const stackId = process.env.STACK_ID!;
+const s3Bucket = process.env.PIPELINE_S3_BUCKET_NAME!;
+let s3Prefix = process.env.PIPELINE_S3_PREFIX!;
+
+if (!s3Prefix.endsWith('/')) {
+  s3Prefix = s3Prefix + '/';
+}
+
+const appIdKey = s3Prefix + stackId + '/config/emr-serverless-applicationId.json';
+
+const emrServerlessClient = new EMRServerlessClient({
+  ...aws_sdk_client_common_config,
+  region,
+});
+
+export const handler = async (event: CloudFormationCustomResourceEvent, context: Context) => {
+  logger.info(JSON.stringify(event));
+  try {
+    const res = await _handler(event, context);
+    logger.info('lambda returned', { res });
+    return res;
+  } catch (e: any) {
+    logger.error(e);
+    throw e;
+  }
+};
+
+async function _handler(event: CloudFormationCustomResourceEvent, context: Context) {
+  logger.info(`functionName: ${context.functionName}`);
+  const props = event.ResourceProperties as ResourcePropertiesType;
+  if (event.RequestType == 'Delete') {
+    await deleteEMRServerlessApp();
+    return;
+  } else {
+    const applicationId = await createEMRServerlessApp(props);
+    return {
+      Data: {
+        ApplicationId: applicationId,
+      },
+    };
+  }
+}
+
+async function createEMRServerlessApp(props: ResourcePropertiesType): Promise<string> {
+  const input: CreateApplicationCommandInput = {
+    name: props.name,
+    releaseLabel: props.version,
+    type: 'SPARK',
+    architecture: Architecture.X86_64,
+    networkConfiguration: {
+      subnetIds: props.subnetIds.split(','),
+      securityGroupIds: [props.secourityGroupId],
+    },
+    autoStartConfiguration: {
+      enabled: true,
+    },
+    autoStopConfiguration: {
+      enabled: true,
+      idleTimeoutMinutes: parseInt(props.idleTimeoutMinutes),
+    },
+  };
+
+  logger.info('CreateApplicationCommand input', { input });
+  const command = new CreateApplicationCommand(input);
+  const response = await emrServerlessClient.send(command);
+  const applicationId = response.applicationId!;
+  logger.info('created emr application applicationId:' + applicationId);
+
+  logger.info('s3Bucket:' + s3Bucket + ', appIdKey:' + appIdKey);
+  let appIdCofing = await readS3ObjectAsJson(s3Bucket, appIdKey);
+  const nowStr = process.env.TEST_TIME_NOW_STR || new Date().toISOString();
+
+  if (appIdCofing) {
+    logger.info('find appIdCofing', { appIdCofing });
+    // only save previous 5 appIds
```

Comment:
Were those apps failed to be deleted?

Reply:
if there are jobs are running, the app will be failed to delete.  if the deletion failed, just ignore the error.  the app will be deleted in next upgrade/update as we saved the previous 5 apps.